Tidy tests

joncotton opened the pull request:
This is a better, cleaner version of this attempted PR #18.

It removes unused imports. It orders them in what I think is a cleaner way (you can tell me to stop this if you just think it's horrible). I found some new things too.
- `TestCase` now uses `ArmstrongTestCase` instead of `DjangoTestCase`
- `TestCase` did setUp() that was only used in `SimpleWellViewTest` so I moved those tasks over
- I removed an unnecessary parent test case
- I found an orphan sanity check

It's meant as an extension to PR #20. So the actual diff is simpler. Here's the [diff](https://github.com/joncotton/armstrong.core.arm_wells/compare/additional_tests...tidy_tests) between those two.

joncotton committed:
Add a test for all() on GFKQuerySet and MergeQuerySet.

joncotton committed:
Clean up _utils.TestCase.

- make it use the base ArmstrongTestCase
- move the setUp() and assertInContext() methods out of the base class and into the class they are used in
- remove an unnecessary WellViewTestCase class

joncotton committed:
Tidy imports. Remove unused imports. Some PEP8 spacing.

joncotton committed:
Include a sanity test that was orphaned.

tswicegood commented on armstrong/core/arm_wells/tests/views.py:
```diff
@@ -1,35 +1,38 @@
             "well_title": self.well.title,
         }
 
+    def assertInContext(self, var_name, other, template_or_context):
```
It's best to keep `assert*` methods like this outside of the test classes they're used.  Invariably, you end up wanting to use them elsewhere if they're useful.  If this is to be moved anywhere, it should be in to `armstrong.dev` so the method is available everywhere.

joncotton replied:
For now should I move it back into `_utils.py.TestCase`?

tswicegood replied:
That's one option, or if you'd like to move this into `armstrong.dev` entirely that's a good option as well.

joncotton replied:
PR [#9](https://github.com/armstrong/armstrong.dev/pull/9)

tswicegood commented on armstrong/core/arm_wells/tests/arm_wells_support/tests.py:
```diff
@@ -1,7 +1,6 @@
 
-from django.test import TestCase
-
+from .._utils import TestCase
```
Tests inside `*_support` modules should not rely on the behavior in any of Armstrong's `TestCase` classes.  These should be straight Django.

tswicegood commented on armstrong/core/arm_wells/tests/managers.py:
```diff
@@ -1,9 +1,7 @@
 import datetime
 
 from ._utils import TestCase
-
```
This line is significant.  Importing in Python is broken down into three groups.  System level, apps that require installation, and finally any specific to this package.  In testing, I try to break them into a fourth -- parts of the system under test.

tswicegood commented on armstrong/core/arm_wells/tests/__init__.py:
```diff
@@ -3,3 +3,4 @@
 from .query import *
 from .querysets import *
 from .views import *
+from .arm_wells_support.tests import *
```
This should not be done -- `arm_wells.tests` is the actual tests of `armstrong.core.arm_wells` while the `arm_wells_support` tests are for the functionality that provides.  They're a sanity check, but should be tested as a separate app.  They won't always be in the `INSTALLED_APPS` which means their models wouldn't be installed, which would cause an issue if `arm_wells` was tested inside another project.

joncotton replied:
That makes sense. Thanks for the explanation. I'll drop the import.

joncotton committed:
Feedback from the PR. Restoring significant import whitespace, using Django TestCase in support test, and leaving that support test out of the package testing.

joncotton committed:
Moving `assertInContext` to armstrong.dev in armstrong.dev PR #9.

tswicegood commented:
The `requirements/dev.txt` file needs to be updated now to ensure that `armstrong.dev>=1.11.0` is installed. Once that's done, I think we're set.

joncotton commented:
halp! Not quite ready. Version conflicts. I'll update requirements.base.
**requirements/base.txt** has `django-reversion==1.3.3`
**package.json** has `"django-reversion==1.4.0"`

joncotton committed:
Updating required versions of `armstrong.dev` and `django-reversion`.

`armstrong.dev` now has the `assertInContext` method.
`django-reversion` should be at 1.4, same as the package.json file and same as armstrong.core.arm_context

joncotton commented:
@tswicegood **_before you merge**_ This includes/builds-off the additional test added in this original PR #20. Merge that one first and this PR will get a little simpler.

tswicegood commented on armstrong/core/arm_wells/tests/querysets.py:
```diff
@@ -178,9 +182,11 @@ def setUp(self):
-    def test_raises_NotImplementedError_on_exclude(self):
+    def test_raises_NotImplementedError_on_all_and_exclude(self):
+        merge_qs = MergeQuerySet(self.qs_a, self.qs_b)
+        with self.assertRaises(NotImplementedError):
```
Can you please break these out into multiple test methods?  If this `assertRaises` failed, we wouldn't know that the next `assertRaises` also fails (or passes) until we fixed the first method.  It's generally a good rule to keep one assertion per test method -- or at least one concern.  Here, `all()` and `exclude()` are two different completely different methods that just happen to raise the same error.

joncotton committed:
Separate tests for `all()` and `exclude()`

tswicegood commented on requirements/base.txt:
```diff
@@ -1,3 +1,3 @@
 django>=1.3
-django-reversion==1.3.3
+django-reversion==1.4
```
Better change.  Let's nix this file entirely and update `requirements/dev.txt` to not include it.

joncotton committed:
Removing the requirements/base.txt file.

niran commented on armstrong/core/arm_wells/tests/querysets.py:
```diff
@@ -1,10 +1,12 @@
+                     add_n_random_images_to_well,
+                     generate_random_image,
+                     generate_random_story,
+                     generate_random_well)
```
I prefer the previous style here, but whatever.

tswicegood replied:
Ditto.  I've reverted some of these in https://github.com/joncotton/armstrong.core.arm_wells/pull/1 to keep it consistent with imports elsewhere.  They should really either be one import, one line; or multi-line imports with `()` like this was originally.

niran commented:
This looks ready to go. There are a lot of different things going on in this PR that were hard to isolate. A PR for adding missing tests (like #22 was, but removed from this one), one for PEP 8 fixes, and one for code clarification would've been easier to sign off on.

joncotton commented:
Well #20 was a separate PR just adding the tests. It didn't get merged by the time I wrote this.

tswicegood commented:
@joncotton sorry that didn't get merged in before this started.  It's best to start all of your feature branches off of `master`, however.  That way you don't have to clean out parts of the PR if the other code doesn't get merged for some reason.  Of course, there is an exception to the rule.  If one PR builds off of code that's in a previous one, starting from that point is what you have to do.
